features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/presequite/SessionReplayRequirement.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.sessionreplay.internal.presequite
+
+import com.datadog.android.api.InternalLogger
+import okhttp3.internal.toImmutableMap
+
+internal class SessionReplayRequirement private constructor(
+    private val checkers: List<SystemRequirementChecker>,
+    private val systemRequirementConfiguration: SystemRequirementConfiguration
+) {
+
+    fun runIfMeetRequirements(
+        featureName: String,
+        internalLogger: InternalLogger,
+        runnable: () -> Unit
+    ) {
+        if (checkAllMinimumRequirements()) {
+            runnable()
+        } else {
+            internalLogger.log(
+                level = InternalLogger.Level.INFO,
+                listOf(InternalLogger.Target.TELEMETRY, InternalLogger.Target.USER),
+                messageBuilder = {
+                    "$featureName is disabled because the system " +
+                        "doesn't meet the minimum SDK requirements"
+                },
+                onlyOnce = true,
+                additionalProperties = getCheckerReport()
+            )
+        }
+    }
+
+    private fun checkAllMinimumRequirements(): Boolean {
+        if (checkers.isEmpty()) {
+            return true
+        }
+        return checkers.all {
+            it.checkMinimumRequirement(systemRequirementConfiguration)
+        }
+    }
+
+    private fun getCheckerReport(): Map<String, Any?> {
+        val map = mutableMapOf<String, Any?>()
+        checkers.forEach { checker ->
+            map[checker.name()] = checker.checkedValue()
+        }
+        return map.toImmutableMap()
+    }
+
+    class Builder {
+        private var checkers = mutableListOf<SystemRequirementChecker>()
+        private var systemRequirementConfiguration: SystemRequirementConfiguration =
+            SystemRequirementConfiguration.DEFAULT
+
+        fun addChecker(checker: SystemRequirementChecker): Builder {
+            checkers.add(checker)
+            return this
+        }
+
+        fun setSystemRequirementConfiguration(systemRequirementConfiguration: SystemRequirementConfiguration): Builder {
```

Comment:
side thought: with such API if I, for example, add requirement configuration, but forget to add checker for the parameter in the configuration, then it will pass even if requirement is not satisfied?

Reply:
yes, my point is rather if it is expected behavior. Normally, if there is a requirement in the config, but no checker for it registered, imo we at least should log that.

Reply:
indeed, I will add a log to `User` if you agree with it.